### PR TITLE
Warn if class has a render() method but doesn't extend React.Component

### PR DESCRIPTION
**Pull request by sw-yx**
Hi! adding the fix for #10103. test/prettier/lint/flow all done and passing. (replaces PR #11149)

### CODE REVIEW

**Comment by gaearon** [comment]
Prettier is failing. Have you run `yarn` before running prettier?

**Comment by sw-yx** [comment]
gosh really embarrassing. Looks like i didnt commit the prettified file. I have resubmitted and it should pass. Jeez I am going to be so embarrassed when I look back on this in a few months

File: src/renderers/shared/fiber/ReactFiberBeginWork.js [hunk 1]
```diff
@@ -481,6 +481,12 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
     var value;
 
     if (__DEV__) {
+      warning(
+        !(workInProgress.type.prototype &&
+          workInProgress.type.prototype.render),
+        'Warning: Using a class with a render method as a function, ' +
```
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Please remove `Warning: `, it gets added automatically.

File: src/renderers/shared/fiber/ReactFiberBeginWork.js [hunk 2]
```diff
@@ -481,6 +481,12 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
     var value;
 
     if (__DEV__) {
+      warning(
+        !(workInProgress.type.prototype &&
```
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: This is a bit hard to follow. Let's just wrap it in an `if` block with the real condition, and then pass `false` to the `warning` call.

File: src/renderers/shared/fiber/ReactFiberBeginWork.js [hunk 3]
```diff
@@ -481,6 +481,12 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
...
+          workInProgress.type.prototype.render),
+        'Warning: Using a class with a render method as a function, ' +
+          'did you forget to extend React.Component?',
```
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: I would tweak the wording to:

>The <%s /> component appears to have a render method, but doesn't extend React.Component. This is likely to cause errors. Change %s to extend React.Component instead.

You’ll need to pass the component name (use `getComponentName`).

**Comment by sw-yx** [comment]
thanks @gaearon . its not clear to me -where- to add the tests. closest i can find is `ReactIncrementalSideEffects-test.js`. is there a deeper test guide somewhere?  I could also put it in `ReactIncremental-test.js` but it is not obvious to me where the natural home is because of lack of similar tests.

**Comment by sw-yx** [comment]
added the test in `ReactIncremental-test.js` - no idea if this is the right place since i couldnt find similar tests to colocate this with but hey let me know...

**Comment by gaearon** [comment]
Sorry the structure is a bit confusing. We haven’t updated the file structure to be sensible yet after deleting the old engine.

But I’d put it [here](https://github.com/facebook/react/blob/master/src/renderers/__tests__/ReactCompositeComponent-test.js).

**Comment by gaearon** [comment]
Also the test is failing:

```

    
    Expected spy to have been called with:
      "Warning: The <ClassWithRenderNotExtended /> component appears to have a render method, but doesn't extend React.Component. This is likely to cause errors. Change ClassWithRenderNotExtended to extend React.Component instead." as argument 1, but it was called with "Warning: The <ClassWithRenderNotExtended /> component appears to have a render method, but doesn't extend React.Component.This is likely to cause errors. Change ClassWithRenderNotExtended to extend React.Component instead.".
```

Have you tried running it with `npm test -- <filename>`?

**Comment by sw-yx** [comment]
yes, i did, and the test passed. seems i errantly deleted a space for no reason, very bizarre. will fix and move as suggested.

**Comment by sw-yx** [comment]
let me know if theres anything else to do!

**Comment by sw-yx** [comment]
let me know if theres anything else to add!

**Comment by gaearon** [comment]
Sorry for the trouble again. We rearranged how files are laid out in the repo. Could you please reapply your changes on top of master?

**Comment by sw-yx** [comment]
I confess I'm not sure what's going on. I thought I could simply `git pull` your changes from master and `git merge` it into my branch, however when the checks run (as they have here) there are -hundreds- of test failures presumably because the files have been moved around. I'm not sure why the tests are not updated in my branch. I'm going to try junking my branch and simply re-creating the branch fresh from your master

**Comment by sw-yx** [comment]
ok this one looks like it works. Woohoo!

File: packages/react-dom/src/__tests__/ReactCompositeComponent-test.js [hunk 4]
```diff
@@ -326,6 +326,32 @@ describe('ReactCompositeComponent', () => {
...
+      class ClassWithRenderNotExtended {
+        render() {
+          return <div />;
+        }
+      }
+      expectDev(console.error.calls.count()).toBe(0);
+      try {
+        ReactDOM.render(<ClassWithRenderNotExtended />, container);
+      } catch (e) {
+        expect(e).toEqual(new TypeError('Cannot call a class as a function'));
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactCompositeComponent-test.js]: This only runs assertions if we get into `catch`. But what if we have a bug that causes us *not* to throw an error? Then the test doesn't assert anything.

**Comment by gaearon** [comment]
I made a few minor tweaks:

* Made sure the test fails if we don't throw.
* Simplified the condition that we check.

I think this is good to go. Thanks!
